#Проект: Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [31]:
children_pivot = data.pivot_table(index = ['children'], values = 'debt', aggfunc = ['count', 'sum', 'mean'])
children_pivot.columns = ['Общее количество клиентов категории', 'Количество невозвратов', '%']
children_pivot['%'] = children_pivot['%'].round(4)*100
children_pivot


Общее количество клиентов категории  Количество невозвратов     %
children                                                                   
0                                       14091                    1063  7.54
1                                        4808                     444  9.23
2                                        2052                     194  9.45
3                                         330                      27  8.18
4                                          41                       4  9.76
5                                           9                       0  0.00

**Вывод:**
___Прямой зависимости нет, если проигнорировать выброс в случае категории клиентов с 3 и 5 детьми, то можно наблюдать рост невозвратов в процентном соотношении, по мере увеличения количества детей. Наиболее надежными заемщиками являются люди без детей или имеющие 3 - х детей, их доля невозвратов самая низкая среди предлагаемых категорий (8.18%). Далее идут клиенты имеющие 1 или 2 детей, их доля невозвратов чуть выше (9.23 и 9.45%). Самые ненадежные заемщики с высокой долей невозвратов (9.76%), это категория клиентов с 4 детьми. В случае с категорией имеющих 5 детей, выборка не показательна, по причине недостаточного количества клиентов для осуществления выводов.___

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [36]:
family_status_pivot = data.pivot_table(index=['family_status'], values='debt',aggfunc=['count','sum', 'mean'])
family_status_pivot.columns = ['Общее количество клиентов категории', 'Количество невозвратов', '%']
family_status_pivot['%'] = family_status_pivot['%'].round(4)*100 
family_status_pivot

Общее количество клиентов категории  \
family_status                                                
Не женат / не замужем                                 2796   
в разводе                                             1189   
вдовец / вдова                                         951   
гражданский брак                                      4134   
женат / замужем                                      12261   

                       Количество невозвратов     %  
family_status                                        
Не женат / не замужем                     273  9.76  
в разводе                                  84  7.06  
вдовец / вдова                             63  6.62  
гражданский брак                          385  9.31  
женат / замужем                           927  7.56

**Вывод:** ___Зависимость есть. Клиенты не состоящие в браке имеют самую высокую долю невозвратов по кредиту (9.76%). Чуть ниже доля (9.31%) невозвратов у клиентов состоящих в гражданском браке. Далее следуют клиенты состоящие в браке (7.56%). Самыми надежными заемщиками являются категории клиентов в разводе (7.06%) и овдовевших супругов (6.62%).___

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [37]:
income_category_pivot = data.pivot_table(index=['total_income_category'], values='debt', aggfunc=['count','sum','mean'])
income_category_pivot.columns = ['Общее количество клиентов категории', 'Количество невозвратов', '%']
income_category_pivot['%'] = income_category_pivot['%'].round(4)*100 
income_category_pivot

Общее количество клиентов категории  \
total_income_category                                        
A                                                       25   
B                                                     5014   
C                                                    15921   
D                                                      349   
E                                                       22   

                       Количество невозвратов     %  
total_income_category                                
A                                           2  8.00  
B                                         354  7.06  
C                                        1353  8.50  
D                                          21  6.02  
E                                           2  9.09

**Вывод:** ___Самая высокая доля (9.09%) просрочек по кредиту у категории 'E' (уровень дохода до 30 тыс.руб), клиенты относящиеся к данной категории, являются по мнению банка, ненадежными заемщиками. Далее по росту надежности идут клиенты категории 'C'- (уровень дохода 50 -200 тыс.руб) (8.50%),'A'- (уровень дохода 1 млн.руб и выше) (8%) и категория 'B' - (уровень дохода 200 тыс.руб - 1 млн.руб) (7.06%). Клиенты категории 'D' (уровень дохода 30 - 50 тыс.руб) являются самыми надежными заемщиками (6.02%). Так же можно заметить, что по мере роста доходов, сначала наблюдается увеличение роста количества невозвратов ('C'- уровень дохода 50 -200 тыс.руб - 8.50%), а затем происходит улучшение динамики в диапозоне клиентов категории B (уровень дохода 200 тыс.руб - 1 млн.руб. - 7.06%). В общем и целом, категории B и D для банка являются самыми желанными клиентами. Исходя из данных показателей, делаю вывод что, прямая зависимость роста количества невозвратов от уровня дохода отсутствует.___

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [38]:
purpose_category_pivot = data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['count', 'sum', 'mean'])
purpose_category_pivot.columns = ['Общее количество клиентов категории', 'Количество невозвратов', '%']
purpose_category_pivot['%'] = purpose_category_pivot['%'].round(4)*100
purpose_category_pivot

Общее количество клиентов категории  \
purpose_category                                                
операции с автомобилем                                   4279   
операции с недвижимостью                                10751   
получение образования                                    3988   
проведение свадьбы                                       2313   

                          Количество невозвратов     %  
purpose_category                                        
операции с автомобилем                       400  9.35  
операции с недвижимостью                     780  7.26  
получение образования                        369  9.25  
проведение свадьбы                           183  7.91

**Вывод:** ___Наиболее надежными заемщиками являются те, кто берут кредиты на операции с недвижимостью (7%). Менее надежными являются клиенты, которые берут кредиты на проведение свадьбы (8%). Самые ненадежные обращаются за кредитом на покупку автомобиля и получения образования (9%).___

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* ___Причинами могут быть ошибки ввода или сокрытие информации.___

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* ___Медианное значение имеет наиболее устойчивую характеристику, так как на нее не влияют случайные выбросы. Если количество данных велико и они имеют нормальное распределение, то медиана будет равняться среднему значению.___

### Шаг 4: общий вывод.


___Наиболее желанным заемщиком для банка является клиент, подходящий под следующие характеристики. Люди состоящие в браке, вдовцы или вдовы, не имеющие детей, с уровнем дохода категории ‘D’ (30 тыс. рублей – 50 тыс. рублей), цель получения кредита: операции с недвижимостью или проведение свадьбы.<br><br>
В противовес, наиболее рискованным клиентом является заемщик не состоящий в официальном браке или живущий в гражданском браке, содержащий 1, 2 или  4 детей, с уровнем дохода категории ‘E’ (до 30 тыс. рублей), цель получения кредита: операции с автомобилем или получение образования.___ 
